ably/realtime/connection.py

```diff
@@ -49,12 +52,20 @@ async def connect(self):
 
     async def close(self):
         self.__state = ConnectionState.CLOSING
-        if self.__websocket:
-            await self.__websocket.close()
+        self.__closed_future = asyncio.Future()
+        if self.__websocket and self.__state == ConnectionState.CONNECTED:
+            task = asyncio.create_task(self.close_connection())
+            await task
         else:
-            log.warn('Connection.closed called while connection already closed')
+            log.warn('Connection.closed called while connection already closed or not established')
         self.__state = ConnectionState.CLOSED
 
+    async def close_connection(self):
```

Comment:
I think the name of this method is slightly unintuitive, perhaps it would be better to call it something like `send_close_message`.